### ISSUE TITLE
vi modal indicator

### ISSUE THREAD

**Issue by shanemikel** [kkawakam/rustyline#208]
Readline supports the vi editing mode, with these options in `~/.inputrc`
```
set editing-mode vi
set show-mode-in-prompt on
set vi-cmd-mode-string :
set vi-ins-mode-string +
```

The `show-mode-in-prompt` feature is a really helpful bit of user feedback for a modal editing system like Vi.  You can test this feature by writing the configuration above, and running Bash.

**Comment by AdminXVII** [comment]
I think a good way would be to follow NVim's way: using a line cursor ('\033[6 q') for insert and emacs mode and switch to a block cursor for command mode.

**Comment by shanemikel** [comment]
I disagree @AdminXVII.  The issue is that not all terminals support cursor modes.

The default behavior of iPython (which doesn't use readline anymore) is to show `[ins]` in insert mode, and `[nav]` in normal/navigation mode.  I think that's pretty good and self explanatory.

**Comment by shanemikel** [comment]
I should also say that the default indicators of readline are `(cmd)` for normal/command mode and `(ins)` for insert mode.  Maybe it's better to use the readline defaults given the library's lineage.

**Comment by max-sixty** [comment]
FWIW this is the config I use for IPython, to get cursors:

```python
   3   │ import sys
   4   │ from prompt_toolkit.key_binding.vi_state import InputMode, ViState
   5   │
   6   │
   7   │ def get_input_mode(self):
   8   │     return self._input_mode
   9   │
  10   │
  11   │ def set_input_mode(self, mode):
  12   │     shape = {InputMode.NAVIGATION: 2, InputMode.REPLACE: 4}.get(mode, 6)
  13   │     raw = u'\x1b[{} q'.format(shape)
  14   │     if hasattr(sys.stdout, '_cli'):
  15   │         out = sys.stdout._cli.output.write_raw
  16   │     else:
  17   │         out = sys.stdout.write
  18   │     out(raw)
  19   │     sys.stdout.flush()
  20   │     self._input_mode = mode
  21   │
  22   │
  23   │ ViState._input_mode = InputMode.INSERT
  24   │ ViState.input_mode = property(get_input_mode, set_input_mode)
  25   │ c.TerminalInteractiveShell.editing_mode = 'vi'
  26   │ c.TerminalInteractiveShell.prompt_includes_vi_mode = False
```

**Comment by pjetherton** [comment]
> The issue is that not all terminals support cursor modes.

Well can't we have a graceful fall-back for those that don't?  Or allow this as a configuration option somewhere?

**Comment by gwenn** [comment]
Just for reference:
 - https://github.com/judah/haskeline/pull/78
 - https://github.com/prompt-toolkit/python-prompt-toolkit/issues/192

**Comment by gwenn** [comment]
For reference, replxx uses a callback to update prompt dynamically:
https://github.com/AmokHuginnsson/replxx/blob/master/examples/cxx-api.cxx#L228-L232
